File: services/elitea-deepwiki/src/elitea_deepwiki/repo_config.py
"""Repository configuration extraction — copied from the legacy handler.

``_extract_repo_config_from_toolkit`` and its four helpers are lifted out of
``deepwiki_plugin/methods/invoke.py`` at revision
``ce679f11dc31c209cc67f13565b286d5bb28ce58``. They are adapter code, not engine
code — they normalise whatever shape the platform's toolkit expansion produced
into the ``repo_config`` dict the engine's tools expect — so they live here
rather than in the copied ``engine/`` package.

They are copied rather than rewritten for the same reason the engine is: 140
lines across four providers (GitHub, GitLab, Bitbucket, Azure DevOps), each
with its own precedence chain over a dozen differently-prefixed keys, several
of which exist only because the UI and the API disagreed about where a value
lives. Retyping that is how a provider quietly stops resolving its branch.
Only the GitHub path is covered by a P0 fixture
(``generation/composed_result.json``'s ``engine_call.repo_config``); the other
three are carried on trust, which is an argument for copying, not against.

TWO MODIFICATIONS, named here because the copy is otherwise verbatim.

1. The legacy code logged its suspicious-ADO warning through Pylon's ``log``.
   That is the stdlib logger below.
2. A FIFTH provider type, ``artifact``, which the legacy plugin never had: a
   wiki whose source is a folder in the invoking project's artifact store
   rather than a git remote. It is a new branch, placed BEFORE the legacy
   four-provider test and returning early, so no legacy path changes shape.
   See ``elitea_deepwiki.artifact_source`` for what the engine then does with
   it, and ``run/repoconfig.go`` for the Go twin of this branch.

Nothing else differs.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, Optional

from .artifact_source import (
    ARTIFACT_SCHEME,
    is_artifact_source,
    parse_artifact_source,
    repo_config_for,
    source_from_configuration,
)

log = logging.getLogger(__name__)


_TOOLKIT_PROVIDER_KEYS = (
    'github_configuration',
    'gitlab_configuration',
    'bitbucket_configuration',
    'ado_configuration',
    # The fifth, added here rather than only in the branch below so that the
    # prefixed and unprefixed spellings merge the same way every other
    # provider's do. The UI and the API disagree about the prefix for all of
    # them, and an artifact source is no exception.
    'artifact_configuration',
)


ARTIFACT_CONFIGURATION_KEYS = (
    'artifact_configuration',
    'toolkit_configuration_artifact_configuration',
)


def _extract_artifact_source(repo_settings: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """The fifth provider: a folder in the project's artifact store.

    Returns the normalised ``repo_config``, or ``None`` when this payload
    names no artifact source — in which case the legacy four-provider chain
    below runs exactly as it always did.

    Two spellings are accepted, because two layers name the source. The
    toolkit carries an ``artifact_configuration`` block with a bucket and an
    optional folder prefix; a request that was already normalised carries the
    ``artifact://`` repository string alone. Both resolve to the same source.
    """
    if not isinstance(repo_settings, dict):
        return None

    branch = (
        repo_settings.get('active_branch')
        or repo_settings.get('toolkit_configuration_active_branch')
        or repo_settings.get('version_label')
        or repo_settings.get('branch')
        or 'main'
    )

    for key in ARTIFACT_CONFIGURATION_KEYS:
        source = source_from_configuration(repo_settings.get(key))
        if source is not None:
            return repo_config_for(source, branch)

    repository = repo_settings.get('repository')
    if is_artifact_source(repository):
        return repo_config_for(parse_artifact_source(repository), branch)
    return None


def _payload_contains_provider_key(params: Dict[str, Any], provider_key: str) -> bool:
    """Check known toolkit payload locations for a provider-specific config key."""
    sources = [
        params,
        params.get('code_toolkit'),
        params.get('toolkit_configuration_code_toolkit'),
        params.get('toolkit_configuration_code_repository'),
        params.get('code_repository'),
    ]

    for source in sources:
        if not isinstance(source, dict):
            continue
        if provider_key in source:
            return True
        for nested in (source.get('settings'), source.get('toolkit_config'), _extract_configuration_parameters(source)):
            if isinstance(nested, dict) and provider_key in nested:
                return True

    return False


def _merge_dicts(*values: Any) -> Dict[str, Any]:
    """Shallow-merge dict values, ignoring non-dicts."""
    merged: Dict[str, Any] = {}
    for value in values:
        if isinstance(value, dict):
            merged.update(value)
    return merged


def _extract_configuration_parameters(source: Dict[str, Any]) -> Dict[str, Any]:
    """Extract toolkit configuration payload from known wrapper shapes."""
    configuration = source.get('configuration')
    if isinstance(configuration, dict) and isinstance(configuration.get('parameters'), dict):
        return configuration.get('parameters') or {}
    return {}


def _merge_provider_configs(settings: Dict[str, Any]) -> Dict[str, Any]:
    """Merge prefixed/unprefixed provider configs that may be split by UI/API shape."""
    merged = dict(settings)
    for provider_key in _TOOLKIT_PROVIDER_KEYS:
        prefixed_key = f'toolkit_configuration_{provider_key}'
        provider_config = _merge_dicts(settings.get(provider_key), settings.get(prefixed_key))
        if provider_config:
            merged[provider_key] = provider_config
            merged[prefixed_key] = provider_config
    return merged


def _merge_toolkit_payload(source: Any) -> Dict[str, Any]:
    """Merge toolkit_config/configuration/settings into one repo-config payload."""
    if not isinstance(source, dict):
        return {}

    wrapper_fields = {key: value for key, value in source.items() if key not in ('settings', 'toolkit_config', 'configuration')}
    sources = (
        wrapper_fields,
        source.get('toolkit_config'),
        _extract_configuration_parameters(source),
        source.get('settings'),
    )
    merged = _merge_dicts(*sources)

    for provider_key in _TOOLKIT_PROVIDER_KEYS:
        prefixed_key = f'toolkit_configuration_{provider_key}'
        provider_config = _merge_dicts(
            *(candidate.get(provider_key) for candidate in sources if isinstance(candidate, dict)),
            *(candidate.get(prefixed_key) for candidate in sources if isinstance(candidate, dict)),
        )
        if provider_config:
            merged[provider_key] = provider_config
            merged[prefixed_key] = provider_config

    return merged



def _extract_repo_config_from_toolkit(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Extract repository configuration from expanded code_toolkit.
    
    Supports: github, gitlab, bitbucket, ado_repos, and — added here — artifact
    
    Returns a normalized repo_config dict with:
        - provider_type: str (github, gitlab, bitbucket, ado_repos)
        - provider_config: dict (the provider-specific configuration)
        - repository: str (repository identifier)
        - branch: str (branch name)
        - project: str or None (for Bitbucket/ADO)
        - is_cloud: bool or None (for Bitbucket)
    """
    code_toolkit = (
        params.get('code_toolkit')
        or params.get('toolkit_configuration_code_toolkit')
        or params.get('toolkit_configuration_code_repository')
        or params.get('code_repository')
        or {}
    )
    
    # Initialize with defaults
    repo_config = {
        'provider_type': 'github',
        'provider_config': {},
        'repository': None,
        'branch': 'main',
        'project': None,
        'is_cloud': None,
    }
    
    repo_settings = {}
    if isinstance(code_toolkit, dict):
        repo_settings = _merge_toolkit_payload(code_toolkit)

    if not repo_settings and isinstance(params, dict):
        repo_settings = _merge_provider_configs(params)

    # ELITEA-PLATFORM ADDITION. The artifact source is tested FIRST and
    # returns early: it shares none of the four providers' key precedence, and
    # a payload that names one names nothing else.
    artifact_config = _extract_artifact_source(repo_settings)
    if artifact_config is None and isinstance(params, dict):
        artifact_config = _extract_artifact_source(params)
    if artifact_config is not None:
        return artifact_config

    if isinstance(repo_settings, dict) and any(
        key in repo_settings
        for key in (
            'github_configuration', 'gitlab_configuration', 'bitbucket_configuration', 'ado_configuration',
            'toolkit_configuration_github_configuration', 'toolkit_configuration_gitlab_configuration',
            'toolkit_configuration_bitbucket_configuration', 'toolkit_configuration_ado_configuration',
        )
    ):
        if 'github_configuration' in repo_settings or 'toolkit_configuration_github_configuration' in repo_settings:
            github_config = repo_settings.get('github_configuration') or repo_settings.get('toolkit_configuration_github_configuration') or {}
            repo_config['provider_type'] = 'github'
            repo_config['provider_config'] = github_config
            repo_config['repository'] = (
                repo_settings.get('repository')
                or repo_settings.get('github_repository')
                or repo_settings.get('toolkit_configuration_github_repository')
            )
            repo_config['branch'] = (
                repo_settings.get('active_branch')
                or repo_settings.get('toolkit_configuration_active_branch')
                or repo_settings.get('base_branch')
                or repo_settings.get('toolkit_configuration_base_branch')
                or repo_settings.get('branch', 'main')
            )
        elif 'gitlab_configuration' in repo_settings or 'toolkit_configuration_gitlab_configuration' in repo_settings:
            gitlab_config = repo_settings.get('gitlab_configuration') or repo_settings.get('toolkit_configuration_gitlab_configuration') or {}
            repo_config['provider_type'] = 'gitlab'
            repo_config['provider_config'] = gitlab_config
            repo_config['repository'] = repo_settings.get('repository') or repo_settings.get('toolkit_configuration_repository')
            repo_config['branch'] = (
                repo_settings.get('branch')
                or repo_settings.get('toolkit_configuration_branch')
                or repo_settings.get('active_branch')
                or repo_settings.get('toolkit_configuration_active_branch')
                or repo_settings.get('base_branch')
                or repo_settings.get('toolkit_configuration_base_branch', 'main')
            )
        elif 'bitbucket_configuration' in repo_settings or 'toolkit_configuration_bitbucket_configuration' in repo_settings:
            bitbucket_config = repo_settings.get('bitbucket_configuration') or repo_settings.get('toolkit_configuration_bitbucket_configuration') or {}
            repo_config['provider_type'] = 'bitbucket'
            repo_config['provider_config'] = bitbucket_config
            repo_config['repository'] = repo_settings.get('repository') or repo_settings.get('toolkit_configuration_repository')
            repo_config['branch'] = (
                repo_settings.get('branch')
                or repo_settings.get('toolkit_configuration_branch')
                or repo_settings.get('active_branch')
                or repo_settings.get('toolkit_configuration_active_branch')
                or repo_settings.get('base_branch')
                or repo_settings.get('toolkit_configuration_base_branch', 'main')
            )
            repo_config['project'] = repo_settings.get('project') or repo_settings.get('toolkit_configuration_project')
            repo_config['is_cloud'] = repo_settings.get('cloud') or repo_settings.get('toolkit_configuration_cloud')
        elif 'ado_configuration' in repo_settings or 'toolkit_configuration_ado_configuration' in repo_settings:
            ado_config = repo_settings.get('ado_configuration') or repo_settings.get('toolkit_configuration_ado_configuration') or {}
            repo_config['provider_type'] = 'ado_repos'
            repo_config['provider_config'] = ado_config
            repo_config['repository'] = (
                repo_settings.get('repository_id')
                or repo_settings.get('toolkit_configuration_repository_id')
                or repo_settings.get('repository')
                or repo_settings.get('toolkit_configuration_repository')
            )
            repo_config['branch'] = (
                repo_settings.get('active_branch')
                or repo_settings.get('toolkit_configuration_active_branch')
                or repo_settings.get('base_branch')
                or repo_settings.get('toolkit_configuration_base_branch')
                or repo_settings.get('branch', 'main')
            )
            repo_config['project'] = ado_config.get('project') or repo_settings.get('project') or repo_settings.get('toolkit_configuration_project')
        else:
            # Fallback: assume GitHub with legacy structure
            repo_config['provider_type'] = 'github'
            repo_config['provider_config'] = repo_settings.get('github_configuration', {})
            repo_config['repository'] = repo_settings.get('repository') or repo_settings.get('github_repository')
            repo_config['branch'] = repo_settings.get('base_branch') or repo_settings.get('active_branch', 'main')
    else:
        # Legacy fallback - direct parameters
        repo_config['provider_type'] = 'github'
        repo_config['provider_config'] = params.get('github_configuration', {})
        repo_config['repository'] = params.get('github_repository')
        repo_config['branch'] = params.get('github_base_branch') or params.get('github_branch', 'main')

    if _payload_contains_provider_key(params, 'ado_configuration') and (
        repo_config.get('provider_type') != 'ado_repos' or not repo_config.get('repository')
    ):
        log.warning(
            'Suspicious ADO repo config extraction: provider_type=%s repository=%s branch=%s project=%s',
            repo_config.get('provider_type'),
            repo_config.get('repository'),
            repo_config.get('branch'),
            repo_config.get('project'),
        )

    return repo_config


__all__ = ["ARTIFACT_SCHEME", "_extract_artifact_source", "_extract_repo_config_from_toolkit"]

File: services/elitea-subapp-host/internal/apps/deepwiki/run/repoconfig.go
package run

import (
	"log/slog"
	"strings"

	"github.com/EliteaAI/elitea-platform/services/elitea-subapp-host/internal/spi"
)

func trimSpace(s string) string       { return strings.TrimSpace(s) }
func hasPrefix(s, prefix string) bool { return strings.HasPrefix(s, prefix) }
func lower(s string) string           { return strings.ToLower(s) }

// The provider keys the toolkit expansion can carry, with or without the
// UI's "toolkit_configuration_" prefix.
var toolkitProviderKeys = []string{
	"github_configuration",
	"gitlab_configuration",
	"bitbucket_configuration",
	"ado_configuration",
	// The fifth, listed here and not only in the branch below so that its
	// prefixed and unprefixed spellings merge the way every other provider's
	// do. The UI and the API disagree about the prefix for all of them.
	"artifact_configuration",
}

// RepoConfig is the normalised repository configuration the engine's tools
// expect — the same keys, in the same order, as the legacy dict, because
// composed_result.json's engine_call.repo_config compares against it.
type RepoConfig struct {
	ProviderType   string         `json:"provider_type"`
	ProviderConfig map[string]any `json:"provider_config"`
	Repository     any            `json:"repository"`
	Branch         any            `json:"branch"`
	Project        any            `json:"project"`
	IsCloud        any            `json:"is_cloud"`
}

// Map is the legacy dict form, for the engine call and for comparisons.
func (r RepoConfig) Map() map[string]any {
	return map[string]any{
		"provider_type":   r.ProviderType,
		"provider_config": r.ProviderConfig,
		"repository":      r.Repository,
		"branch":          r.Branch,
		"project":         r.Project,
		"is_cloud":        r.IsCloud,
	}
}

// RepositoryString is the repository as a string, "" when unset.
func (r RepoConfig) RepositoryString() string { return str(r.Repository) }

// BranchString is the branch as a string, "" when unset.
func (r RepoConfig) BranchString() string { return str(r.Branch) }

func mergeDicts(values ...any) map[string]any {
	merged := map[string]any{}
	for _, value := range values {
		for k, v := range object(value) {
			merged[k] = v
		}
	}
	return merged
}

func configurationParametersOf(source map[string]any) map[string]any {
	if p := object(object(source["configuration"])["parameters"]); p != nil {
		return p
	}
	return map[string]any{}
}

func payloadContainsProviderKey(params map[string]any, providerKey string) bool {
	sources := []any{
		params,
		params["code_toolkit"],
		params["toolkit_configuration_code_toolkit"],
		params["toolkit_configuration_code_repository"],
		params["code_repository"],
	}
	for _, source := range sources {
		m := object(source)
		if m == nil {
			continue
		}
		if _, ok := m[providerKey]; ok {
			return true
		}
		for _, nested := range []map[string]any{object(m["settings"]), object(m["toolkit_config"]), configurationParametersOf(m)} {
			if nested == nil {
				continue
			}
			if _, ok := nested[providerKey]; ok {
				return true
			}
		}
	}
	return false
}

func mergeProviderConfigs(settings map[string]any) map[string]any {
	merged := map[string]any{}
	for k, v := range settings {
		merged[k] = v
	}
	for _, providerKey := range toolkitProviderKeys {
		prefixed := "toolkit_configuration_" + providerKey
		providerConfig := mergeDicts(settings[providerKey], settings[prefixed])
		if len(providerConfig) > 0 {
			merged[providerKey] = providerConfig
			merged[prefixed] = providerConfig
		}
	}
	return merged
}

func mergeToolkitPayload(source any) map[string]any {
	m := object(source)
	if m == nil {
		return map[string]any{}
	}
	wrapper := map[string]any{}
	for k, v := range m {
		if k != "settings" && k != "toolkit_config" && k != "configuration" {
			wrapper[k] = v
		}
	}
	sources := []any{wrapper, m["toolkit_config"], configurationParametersOf(m), m["settings"]}
	merged := mergeDicts(sources...)
	for _, providerKey := range toolkitProviderKeys {
		prefixed := "toolkit_configuration_" + providerKey
		var candidates []any
		for _, candidate := range sources {
			if c := object(candidate); c != nil {
				candidates = append(candidates, c[providerKey])
			}
		}
		for _, candidate := range sources {
			if c := object(candidate); c != nil {
				candidates = append(candidates, c[prefixed])
			}
		}
		providerConfig := mergeDicts(candidates...)
		if len(providerConfig) > 0 {
			merged[providerKey] = providerConfig
			merged[prefixed] = providerConfig
		}
	}
	return merged
}

func hasAny(m map[string]any, keys ...string) bool {
	for _, key := range keys {
		if _, ok := m[key]; ok {
			return true
		}
	}
	return false
}

func orObject(values ...any) map[string]any {
	for _, v := range values {
		if m := object(v); len(m) > 0 {
			return m
		}
	}
	return map[string]any{}
}

// ExtractRepoConfig normalises whatever shape the platform's toolkit
// expansion produced into the repo_config the engine's tools expect. Lifted
// from the legacy invoke.py at ce679f11 by way of the Python shell's
// repo_config.py: four providers, each with its own precedence chain over a
// dozen differently-prefixed keys. Copied rather than rewritten, because
// retyping that is how a provider quietly stops resolving its branch. Only
// the GitHub path is pinned by a fixture; the other three are carried on
// trust, as the Python port's README says.
func ExtractRepoConfig(params map[string]any) RepoConfig {
	codeToolkit := firstTruthy(
		params["code_toolkit"],
		params["toolkit_configuration_code_toolkit"],
		params["toolkit_configuration_code_repository"],
		params["code_repository"],
		map[string]any{},
	)
	config := RepoConfig{ProviderType: "github", ProviderConfig: map[string]any{}, Branch: "main"}

	repoSettings := map[string]any{}
	if ct := object(codeToolkit); ct != nil {
		repoSettings = mergeToolkitPayload(ct)
	}
	if len(repoSettings) == 0 && params != nil {
		repoSettings = mergeProviderConfigs(params)
	}
	s := repoSettings
	// The ARTIFACT source is tested first and returns early: it shares none of
	// the four providers' key precedence, and a payload naming one names none
	// of the others. Everything below this is the legacy chain, unchanged.
	if source, ok := ArtifactSourceOf(s); ok {
		return artifactRepoConfig(source, firstTruthy(
			s["active_branch"], s["toolkit_configuration_active_branch"],
			s["version_label"], s["branch"], "main"))
	}
	if source, ok := ArtifactSourceOf(params); ok {
		return artifactRepoConfig(source, firstTruthy(
			params["active_branch"], params["version_label"], params["branch"], "main"))
	}
	switch {
	case hasAny(s, "github_configuration", "gitlab_configuration", "bitbucket_configuration", "ado_configuration",
		"toolkit_configuration_github_configuration", "toolkit_configuration_gitlab_configuration",
		"toolkit_configuration_bitbucket_configuration", "toolkit_configuration_ado_configuration"):
		switch {
		case hasAny(s, "github_configuration", "toolkit_configuration_github_configuration"):
			config.ProviderType = "github"
			config.ProviderConfig = orObject(s["github_configuration"], s["toolkit_configuration_github_configuration"])
			config.Repository = firstTruthy(s["repository"], s["github_repository"], s["toolkit_configuration_github_repository"])
			config.Branch = firstTruthy(s["active_branch"], s["toolkit_configuration_active_branch"], s["base_branch"], s["toolkit_configuration_base_branch"], get(s, "branch", "main"))
		case hasAny(s, "gitlab_configuration", "toolkit_configuration_gitlab_configuration"):
			config.ProviderType = "gitlab"
			config.ProviderConfig = orObject(s["gitlab_configuration"], s["toolkit_configuration_gitlab_configuration"])
			config.Repository = firstTruthy(s["repository"], s["toolkit_configuration_repository"])
			config.Branch = firstTruthy(s["branch"], s["toolkit_configuration_branch"], s["active_branch"], s["toolkit_configuration_active_branch"], s["base_branch"], get(s, "toolkit_configuration_base_branch", "main"))
		case hasAny(s, "bitbucket_configuration", "toolkit_configuration_bitbucket_configuration"):
			config.ProviderType = "bitbucket"
			config.ProviderConfig = orObject(s["bitbucket_configuration"], s["toolkit_configuration_bitbucket_configuration"])
			config.Repository = firstTruthy(s["repository"], s["toolkit_configuration_repository"])
			config.Branch = firstTruthy(s["branch"], s["toolkit_configuration_branch"], s["active_branch"], s["toolkit_configuration_active_branch"], s["base_branch"], get(s, "toolkit_configuration_base_branch", "main"))
			config.Project = firstTruthy(s["project"], s["toolkit_configuration_project"])
			config.IsCloud = firstTruthy(s["cloud"], s["toolkit_configuration_cloud"])
		default: // ado
			ado := orObject(s["ado_configuration"], s["toolkit_configuration_ado_configuration"])
			config.ProviderType = "ado_repos"
			config.ProviderConfig = ado
			config.Repository = firstTruthy(s["repository_id"], s["toolkit_configuration_repository_id"], s["repository"], s["toolkit_configuration_repository"])
			config.Branch = firstTruthy(s["active_branch"], s["toolkit_configuration_active_branch"], s["base_branch"], s["toolkit_configuration_base_branch"], get(s, "branch", "main"))
			config.Project = firstTruthy(ado["project"], s["project"], s["toolkit_configuration_project"])
		}
	case len(s) > 0 || object(codeToolkit) != nil:
		// The legacy `else` on the provider-key test: settings without any
		// provider key.
		config.ProviderType = "github"
		config.ProviderConfig = orObject(s["github_configuration"])
		config.Repository = firstTruthy(s["repository"], s["github_repository"])
		config.Branch = firstTruthy(s["base_branch"], get(s, "active_branch", "main"))
	}
	if len(s) == 0 && object(codeToolkit) == nil {
		config.ProviderType = "github"
		config.ProviderConfig = orObject(params["github_configuration"])
		config.Repository = params["github_repository"]
		config.Branch = firstTruthy(params["github_base_branch"], get(params, "github_branch", "main"))
	}
	if payloadContainsProviderKey(params, "ado_configuration") && (config.ProviderType != "ado_repos" || !Truthy(config.Repository)) {
		slog.Warn("suspicious ADO repo config extraction",
			"provider_type", config.ProviderType, "repository", config.Repository, "branch", config.Branch, "project", config.Project)
	}
	return config
}

// The provider-config keys that can carry a host, in the legacy order, and
// each provider's public host when the configuration names none.
var (
	hostKeys     = []string{"base_url", "url", "api_url", "host"}
	defaultHosts = map[string]string{"github": "github.com", "gitlab": "gitlab.com", "bitbucket": "bitbucket.org", "ado_repos": "dev.azure.com"}
)

// DestinationHost mirrors what the engine's provider factory does with the
// same dict: the provider's configured base URL, else the repository when it
// is a full URL, else the provider's public default.
func DestinationHost(config RepoConfig) string {
	provider := lower(firstTruthy(config.ProviderType, "github").(string))
	for _, key := range hostKeys {
		if host := spi.HostOf(str(config.ProviderConfig[key])); host != "" {
			return host
		}
	}
	if host := spi.HostOf(config.RepositoryString()); host != "" {
		return host
	}
	return defaultHosts[provider]
}

// CheckEgress refuses a clone destination that is not on the allowlist. A
// request that names no repository at all is not a clone and passes; one
// that names one is checked, and an unresolvable host is refused too. The
// refusal is invalid_input, which the caller already knows how to read.
func CheckEgress(policy spi.EgressPolicy, params map[string]any) (string, error) {
	config := ExtractRepoConfig(params)
	// An ARTIFACT FOLDER reaches no external host, so the allowlist has
	// nothing to decide. This is NOT a hole where the check used to be: what
	// the allowlist protects against is a clone from a host the deployment
	// does not trust, and a folder is read from elitea-main over the callback
	// grant this invocation was minted with. That grant carries the project,
	// so the ONLY buckets reachable are the caller's own project's — the
	// scoping the allowlist would otherwise have to stand in for.
	if config.ProviderType == ArtifactProviderType {
		return "", nil
	}
	if len(config.ProviderConfig) == 0 && !Truthy(config.Repository) {
		return "", nil
	}
	host := DestinationHost(config)
	checked := host
	if checked == "" {
		checked = "<unresolvable>"
	}
	if err := policy.Check(checked, "clone destination"); err != nil {
		return host, err
	}
	return host, nil
}
